### PR TITLE
feat: add git-refs datasource

### LINKED ISSUE #5605: Add `git-refs` datasource
**What would you like Renovate to be able to do?**

Support a full git "refs" datasource in addition to the current `git-tags` datasource.

**Describe the solution you'd like**

Pretty much exactly like `git-tags` but not filtering for only tags: https://github.com/renovatebot/renovate/blob/aaa7728264c01a0c2752e3a4afc7dae52b7f7933/lib/datasource/git-tags/index.ts#L27

Naturally the code should be shared and not duplicated.

**Describe alternatives you've considered**

Today we use the `git-tags` datasource in some places where `git-refs` would be more accurate. i.e. taking into account commits and branch names.

### PR DESCRIPTION
**Pull request by ViralRuparel**
Adds feature to support full git-refs. 

Closes #5605

### CODE REVIEW

**Comment by ViralRuparel** [comment]
Ready for review @rarkins

File: lib/datasource/git-refs/index.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import { logger } from '../../logger';
+import { ReleaseResult, GetReleasesConfig } from '../common';
+
+export const id = 'git-refs';
+
+const cacheNamespace = 'git-refs';
+const cacheMinutes = 10;
+
+// git will prompt for known hosts or passwords, unless we activate BatchMode
+process.env.GIT_SSH_COMMAND = 'ssh -o BatchMode=yes';
```
Comment (viceice) [inline lib/datasource/git-refs/index.ts]: This should be moved to a more general file.

@rarkins @JamieMagee  Any opinion?

File: lib/datasource/git-tags/index.ts [hunk 2]
```diff
@@ -24,16 +23,15 @@ export async function getPkgReleases({
...
+
     // extract valid tags from git ls-remote which looks like 'commithash\trefs/tags/1.2.3
-    const tags = lsRemote
-      .replace(/^.+?refs\/tags\//gm, '')
-      .split('\n')
+    const refs = lsRemote.releases.map(release => release.gitRef);
+
+    const tags = refs
+      .map(ref => ref.replace(/^tags\//gm, ''))
       .filter(tag => semver.isVersion(tag));
```
Comment (viceice) [inline lib/datasource/git-tags/index.ts]: We should not filter by semver here. It should allow whatever versioning scheme the repo uses.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/git-tags/index.ts (hunk 2 above)
Comment (ViralRuparel) [reply lib/datasource/git-tags/index.ts]: It was there already, I did not add it, shall I remove that?

File: lib/datasource/git-refs/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/datasource/git-refs/index.ts]: It will have global effect, so maybe we should do it during global init?

Comment (ViralRuparel) [reply lib/datasource/git-refs/index.ts]: You mean into constants?

File: lib/datasource/git-tags/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/datasource/git-tags/index.ts]: Yes, it might need removing from the git tags datasource this was copy/pasted from too - but not in this PR

Comment (ViralRuparel) [reply lib/datasource/git-tags/index.ts]: it is in git-tags only and not in git-refs!

Comment (viceice) [reply lib/datasource/git-tags/index.ts]: @ViralRuparel Yes, but we should remove that in another pr.

Comment (ViralRuparel) [reply lib/datasource/git-tags/index.ts]: okay @viceice

File: lib/datasource/git-refs/index.ts (hunk 1 above)
Comment (viceice) [reply lib/datasource/git-refs/index.ts]: no, i think here:
https://github.com/renovatebot/renovate/blob/1d661b909b18b5a09dc4b099c72cbd5397a6bb57/lib/workers/global/index.ts#L63-L69

And we shouldn't override if already set.

Comment (ViralRuparel) [reply lib/datasource/git-refs/index.ts]: okay but in another PR right? @viceice

Comment (JamieMagee) [reply lib/datasource/git-refs/index.ts]: A bit late to the party, but I think moving this from `git-refs` and `git-tags` is the best option. You can do it now, or in another PR. I have no preference.

This did send me down a bit of a rabbit hole about what args we should set for git :smile:

File: lib/datasource/git-refs/index.spec.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import _simpleGit from 'simple-git/promise';
+import { getPkgReleases } from '.';
+
+jest.mock('simple-git/promise');
+const simpleGit: any = _simpleGit;
+
+// const lookupName = 'vapor';
```
Comment (JamieMagee) [inline lib/datasource/git-refs/index.spec.ts]: Leftover from debugging?

File: lib/datasource/git-refs/index.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      releases: refs.map(ref => ({
+        version: ref,
+        gitRef: ref,
+      })),
+    };
+
+    await renovateCache.set(cacheNamespace, lookupName, result, cacheMinutes);
+    return result;
+  } catch (e) {
+    logger.debug(`Error looking up refs in ${lookupName}`);
```
Comment (JamieMagee) [inline lib/datasource/git-refs/index.ts]: Feels like more of a warning to me? I know this is also how `git-tags` and `git-submodule` do it, so feel free to change it or leave it as-is and I'll clean it up later.

File: lib/datasource/git-refs/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/datasource/git-refs/index.ts]: @JamieMagee do you mean moving the `process.env` part? And why in `git-tags`?

Comment (JamieMagee) [reply lib/datasource/git-refs/index.ts]: @rarkins Yes, moving `process.env`. And this same line of code also exists in `git-tags`. We should set it once during initialisation.

File: lib/datasource/git-refs/index.ts (hunk 4 above)
Comment (ViralRuparel) [reply lib/datasource/git-refs/index.ts]: okay 👍

File: lib/datasource/git-refs/index.spec.ts (hunk 3 above)
Comment (ViralRuparel) [reply lib/datasource/git-refs/index.spec.ts]: removed

File: lib/datasource/git-refs/index.ts (hunk 4 above)
Comment (ViralRuparel) [reply lib/datasource/git-refs/index.ts]: Done, changed in git-ref, git-tags and git-submodules @JamieMagee

File: lib/datasource/git-refs/index.ts (hunk 1 above)
Comment (ViralRuparel) [reply lib/datasource/git-refs/index.ts]: we can keep it in `git-refs` and remove it from `git-tags` as `git-tags` is now using `git-refs` code and also `git-refs` executes first @JamieMagee

**Review by rarkins — CHANGES_REQUESTED** [review]
I think we have a problem here. My aim was that these would be [committish](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefcommit-ishacommit-ishalsocommittish) references - namely including not just tags but also branch names. If the values are like `tags/0.0.1` then they won't be sortable/filterable/etc.

@viceice @JamieMagee any thoughts here?

**Comment by viceice** [comment]
something like this (the right part)?:
```
❯ git ls-remote --ref
From https://github.com/viceice/bitbucket-branch-source-plugin.git
f242e60a14b426c6f1d09ba45dea314d13fb104b        refs/heads/fix/bbs-7.x
3dbf6b60550d90ab6e98991f51740c4bc279ffb4        refs/heads/master
3dbf6b60550d90ab6e98991f51740c4bc279ffb4        refs/pull/1/head
2f93ef7f6254da528af5b22409309570c2bb3172        refs/tags/bitbucket-branch-source-1.3
0f2777ad602fedae9971745e9eb14bf307b66da5        refs/tags/bitbucket-branch-source-1.4
bafe0a4d46d9d72cf8f864e002f7b353e68737d1        refs/tags/bitbucket-branch-source-1.5
89ba6ac2fc3891af1ad5e3ad567dbf86a617921b        refs/tags/bitbucket-branch-source-1.6
0e8b13f277c054b822597ad2c2b7ba8cdd2ef20c        refs/tags/bitbucket-branch-source-1.7
b4a3128f73f20604a5f99dc3d095dfa16d4da40d        refs/tags/bitbucket-branch-source-1.8
77192705a64b2356e0456d4c06297a1f58b4da7c        refs/tags/bitbucket-branch-source-1.9
265809cf834e9bfdb7d108c3a4eee25e4a98e7ba        refs/tags/bitbucket-branch-source-2.0.0
73f8741fc4eff9fb2f71154fdc51947e49e09f49        refs/tags/bitbucket-branch-source-2.0.0-beta-1
515ec42cda734528b9920784eb979a0fc5d42c67        refs/tags/bitbucket-branch-source-2.0.1
```

**Comment by ViralRuparel** [comment]
> something like this (the right part)?:
> 
> ```
> ❯ git ls-remote --ref
> From https://github.com/viceice/bitbucket-branch-source-plugin.git
> f242e60a14b426c6f1d09ba45dea314d13fb104b        refs/heads/fix/bbs-7.x
> 3dbf6b60550d90ab6e98991f51740c4bc279ffb4        refs/heads/master
> 3dbf6b60550d90ab6e98991f51740c4bc279ffb4        refs/pull/1/head
> 2f93ef7f6254da528af5b22409309570c2bb3172        refs/tags/bitbucket-branch-source-1.3
> 0f2777ad602fedae9971745e9eb14bf307b66da5        refs/tags/bitbucket-branch-source-1.4
> bafe0a4d46d9d72cf8f864e002f7b353e68737d1        refs/tags/bitbucket-branch-source-1.5
> 89ba6ac2fc3891af1ad5e3ad567dbf86a617921b        refs/tags/bitbucket-branch-source-1.6
> 0e8b13f277c054b822597ad2c2b7ba8cdd2ef20c        refs/tags/bitbucket-branch-source-1.7
> b4a3128f73f20604a5f99dc3d095dfa16d4da40d        refs/tags/bitbucket-branch-source-1.8
> 77192705a64b2356e0456d4c06297a1f58b4da7c        refs/tags/bitbucket-branch-source-1.9
> 265809cf834e9bfdb7d108c3a4eee25e4a98e7ba        refs/tags/bitbucket-branch-source-2.0.0
> 73f8741fc4eff9fb2f71154fdc51947e49e09f49        refs/tags/bitbucket-branch-source-2.0.0-beta-1
> 515ec42cda734528b9920784eb979a0fc5d42c67        refs/tags/bitbucket-branch-source-2.0.1
> ```

I think the behavior of the current code is exactly the same as here.

**Comment by ViralRuparel** [comment]
I returns all the refs and then I am removing `ref/` from the path and then creating an array!
Just that any other thing other than tags are not included in the test cases, so it is returning just the` tags` as ref. Let me add some more things other than `tags` and check in the test cases!

**Comment by viceice** [comment]
🤔 without `heads` and `tags` prefix we don't know what ref type it is. 🤔

**Comment by ViralRuparel** [comment]
> 🤔 without `heads` and `tags` prefix we don't know what ref type it is. 🤔

Currently, it does have the prefixes like `tags` just removing prefix `refs/` from path!

**Comment by viceice** [comment]
I know, sorry just some thinking. @rarkins do you have sample how it could look like?

**Comment by rarkins** [comment]
My idea was the same process tools like `npm` use. e.g. it might be a reference to `renovatebot/renovate#v20` and that's referring to the `v20` branch, not tag. Similarly, it could be `renovatebot/renovate#v19.165.0` and that's a tag. Maybe `git-refs` just needs to be tags and branches for now?

BTW we've had cases where someone has a committish reference to a beta branch, e.g. `v2.0.0`, while their repo had tags for all the v1.x. Renovate can't find `v2.0.0` tag and so generates a rollback PR. By adding git-refs, I wanted to avoid this.

**Comment by viceice** [comment]
Ok, so maybe we should have two different datasources. One for tags and one for branches.

**Comment by rarkins** [comment]
I don’t think it will work because we don’t know which it is beforehand. Hence this combined datasource.

**Comment by viceice** [comment]
But how to differentiate between branches and tags. 🤔

**Comment by JamieMagee** [comment]
Well you can't without the full ref. `refs/heads/v2` and `refs/tags/v2` both look the same. But does it really matter? If you're doing this, then a lot of normal git commands already break, or get more difficult to use (what does `git diff v2 master` do?)

**Comment by rarkins** [comment]
It's allowed, but chaotic and not recommended. As per https://softwareengineering.stackexchange.com/a/230440, branch names would take preference, so I suggest we do the same

**Comment by rarkins** [comment]
So here's what I'd like:
- Combine branches and tags
- Strip the prefixes
- Deduplicate

File: lib/datasource/common.ts [hunk 5]
```diff
@@ -5,6 +5,7 @@ export interface GetReleasesConfig {
   registryUrls?: string[];
   compatibility?: Record<string, string>;
   npmrc?: string;
+  filterByTags?: string;
```
Comment (rarkins) [inline lib/datasource/common.ts]:
```suggestion
```

We don't want to change this interface

File: lib/datasource/git-refs/index.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+const cacheNamespace = 'git-refs';
+const cacheMinutes = 10;
+
+// git will prompt for known hosts or passwords, unless we activate BatchMode
+process.env.GIT_SSH_COMMAND = 'ssh -o BatchMode=yes';
+
+export async function getPkgReleases({
+  lookupName,
+  filterByTags,
```
Comment (rarkins) [inline lib/datasource/git-refs/index.ts]:
```suggestion
```

No change to interface

**Review by rarkins — CHANGES_REQUESTED** [review]
For sharing code between git-refs and git-tags, please export a different function from git-refs to git-tags. Additionally, git-refs should not be "aware" of git-tags. This function can simply return the results including *type* and let git-tags and git-refs filter them differently.

File: lib/datasource/git-tags/index.ts [hunk 7]
```diff
@@ -1,20 +1,16 @@
-import simpleGit from 'simple-git/promise';
+import { ReleaseResult, GetReleasesConfig } from '../common';
 import * as semver from '../../versioning/semver';
 import { logger } from '../../logger';
-import { ReleaseResult, GetReleasesConfig } from '../common';
+import * as gitRefs from '../git-refs';
 
 export const id = 'git-tags';
 
 const cacheNamespace = 'git-tags';
```
Comment (rarkins) [inline lib/datasource/git-tags/index.ts]: I don't think we need to cache this anymore if we're caching the getRawRefs() result, right?

Comment (ViralRuparel) [reply lib/datasource/git-tags/index.ts]: yeah, that's right ! but what I thought was, still if this is cached and then we can even save the process of filtering the tags from it for that much time!, what do you think shall I remove git-refs and git-tags cache and only cache the getRawRefs() results?

Comment (rarkins) [reply lib/datasource/git-tags/index.ts]: Yes, remove both cases and only cache the getRawRefs results

File: lib/datasource/common.ts [hunk 8]
```diff
@@ -49,6 +49,10 @@ export interface ReleaseResult {
   versions?: any;
 }
 
+export interface RawRefs {
```
Comment (rarkins) [inline lib/datasource/common.ts]: Please move this to the `git-refs` file too.

Comment (ViralRuparel) [reply lib/datasource/common.ts]: okay

File: lib/datasource/git-tags/index.ts (hunk 7 above)
Comment (ViralRuparel) [reply lib/datasource/git-tags/index.ts]: okay

File: lib/datasource/git-refs/index.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import { ReleaseResult, GetReleasesConfig } from '../common';
+
+export const id = 'git-refs';
+
+const cacheMinutes = 10;
+
+// git will prompt for known hosts or passwords, unless we activate BatchMode
+process.env.GIT_SSH_COMMAND = 'ssh -o BatchMode=yes';
+
+interface RawRefs {
```
Comment (rarkins) [inline lib/datasource/git-refs/index.ts]:
```suggestion
export interface RawRefs {
```

File: lib/datasource/git-refs/index.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    logger.debug({ err }, `Git-Raw-Refs lookup error in ${lookupName}`);
+  }
+  return null;
+}
+
+export async function getPkgReleases({
+  lookupName,
+}: GetReleasesConfig): Promise<ReleaseResult | null> {
+  try {
+    const rawRefs = await getRawRefs({ lookupName });
```
Comment (rarkins) [inline lib/datasource/git-refs/index.ts]:
```suggestion
    const rawRefs: RawRefs[] = await getRawRefs({ lookupName });
```

File: lib/datasource/git-tags/index.ts [hunk 11]
```diff
@@ -1,42 +1,24 @@
...
-    const lsRemote = await git.listRemote([
-      '--sort=-v:refname',
-      '--tags',
-      lookupName,
-    ]);
-    // extract valid tags from git ls-remote which looks like 'commithash\trefs/tags/1.2.3
-    const tags = lsRemote
-      .replace(/^.+?refs\/tags\//gm, '')
-      .split('\n')
+    const rawRefs = await gitRefs.getRawRefs({ lookupName });
```
Comment (rarkins) [inline lib/datasource/git-tags/index.ts]:
```suggestion
    const rawRefs: gitRefs.RawRefs[] = await gitRefs.getRawRefs({ lookupName });
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Pass type definitions through

File: lib/datasource/git-tags/index.ts (hunk 11 above)
Comment (ViralRuparel) [reply lib/datasource/git-tags/index.ts]: okay gotcha!

**Review by viceice — APPROVED** [review]

**Pull request merged**